pkg/api/middlewares.go

```diff
@@ -27,3 +29,52 @@ func getHTTPCacheGenNumberHeaderSetterMiddleware(cacheGenNumbersLoader *purger.T
 		})
 	})
 }
+
+// HTTPHeaderMiddleware adds specified HTTPHeaders to the request context
+type HTTPHeaderMiddleware struct {
+	TargetHeaders []string
+}
+
+// NewHttpHeaderMiddleware creates a new hTTPHeaderMiddleware given a string containing headers delimited by ~ (whitespaces in headers ignored)
+func NewHttpHeaderMiddleware(unparsedTargetHeaders string) HTTPHeaderMiddleware {
+	removeWhitespaceRegex, err := regexp.Compile(`\s`)
+	if err != nil {
+		util_log.Logger.Log("WhitespaceRegexCompilationStatus", "Failed")
+		return HTTPHeaderMiddleware{TargetHeaders: nil}
+	}
+	headersWithNoWhitespace := removeWhitespaceRegex.ReplaceAllString(unparsedTargetHeaders, "")
+
+	splitRegex, err2 := regexp.Compile(`~`)
+	if err2 != nil {
+		util_log.Logger.Log("SplitRegexCompilationStatus", "Failed")
+		return HTTPHeaderMiddleware{TargetHeaders: nil}
+	}
+	splitHeaders := splitRegex.Split(headersWithNoWhitespace, -1)
+	return HTTPHeaderMiddleware{TargetHeaders: splitHeaders}
+
+}
+
+// InjectRequestIdIntoHTTPRequest injects specified HTTPHeaders into the request context
+func (h HTTPHeaderMiddleware) InjectRequestIdIntoHTTPRequest(r *http.Request) context.Context {
+	// Check to make sure that Headers have not already been injected
+	testing, ok := r.Context().Value(util_log.RequestTargetsContextKey).([]string)
+	if ok && h.TargetHeaders != nil && testing[0] == h.TargetHeaders[0] {
+		return r.Context()
+	}
+	headerContents := make([]string, len(h.TargetHeaders))
+	for index, target := range h.TargetHeaders {
+		headerContents[index] = r.Header.Get(target)
```

Comment:
What happen if we dont find the header?
We cant assume that all header will be in all calls

Reply:
Right now it results in an empty string. So an example of the log output would be requestID="" . Would it be better to avoid the clutter and just not have it in the logs, or to keep it/replace with something like "unknown" ?

Reply:
My suggestion is to not have in the logs. This highly depends on the customer calling the API. If they dont send the data i would just ignore it.